amethyst_network/src/server/host.rs

```diff
@@ -1,61 +1,50 @@
-/// This module is about the communication with this host to some other endpoints.
-///
-/// 1. Sending Data
-/// 2. Receiving Data
-/// 3. Broadcasting
-use crate::error::Result;
-use crate::server::{
-    ReceiveHandler, SendHandler, ServerConfig, ServerSocketEvent, UdpReceiver, UdpSender,
-};
-use laminar::Packet;
-use std::sync::{Arc, Mutex};
-
-/// 'Host' abstracts TCP and UDP sockets away.
+//! This module is about the communication with this host to some other endpoints.
+//!
+//! 1. Sending Data
+//! 2. Receiving Data
+//! 3. Broadcasting
+
+use crate::{error::Result, server::ServerConfig};
+use crossbeam_channel::{Receiver, Sender};
+use laminar::{Packet, Socket, SocketEvent};
+use std::thread;
+
+/// 'Host' abstracts Laminar udp sockets away.
 pub struct Host {
-    // Handler to access the internals of the UDP receiving thread
-    udp_receiver: Arc<Mutex<ReceiveHandler>>,
-    // Handler to access the internals of the UDP sender thread
-    udp_sender: Arc<SendHandler>,
+    packet_sender: Sender<Packet>,
+    packet_receiver: Receiver<SocketEvent>,
 }
 
 impl Host {
     /// This will start and return an instance of the host.
-    /// 1: Fire up a TCP-sender and TCP-receiver if enabled in config.
-    /// 2: Fire up a UDP-sender and UDP-receiver.
-    /// 3: Set up some `channels` to communicate with underlying threads.
     ///
     /// The method uses the config provided when creating a `host` instance.
     pub fn run(config: &ServerConfig) -> Result<Host> {
-        // setup a UDP-receiver which will receive packets from any endpoint.
-        let udp_receiver = Arc::new(Mutex::new(UdpReceiver::run(config.udp_recv_addr, &config)?));
+        let (mut socket, packet_sender, packet_receiver) = Socket::bind(config.udp_socket_addr)?;
 
-        // setup the UDP-sender which will send packets to an certain endpoint.
-        let udp_sender = Arc::new(UdpSender::run(config.udp_send_addr)?);
+        thread::spawn(move || {
+            socket.start_polling();
+        });
 
         Ok(Host {
-            udp_sender,
-            udp_receiver,
+            packet_sender,
+            packet_receiver,
         })
     }
 
     /// Get the handle to the internals of the UDP-receiving threat.
-    pub fn udp_receive_handle(&self) -> Arc<Mutex<ReceiveHandler>> {
-        self.udp_receiver.clone()
+    pub fn udp_receive_handle(&self) -> Receiver<SocketEvent> {
```

Comment:
just dropping this here to think about in the future but we probably do not want to expose crossbeam directly but instead wrap it in our own handle struct, this way if crossbeam does decide to make a breaking change we are safe.

Reply:
True, I am a bit searching for a correct implementation but it isn't public so we can change it without causing any problems.